presto-main/src/main/java/io/prestosql/sql/planner/QueryPlanner.java

```diff
@@ -470,6 +475,104 @@ public DeleteNode plan(Delete node)
+        }
+        List<Expression> orderedExpressions = orderedExpressionBuilder.build();
+
+        // create table scan
+        RelationPlan relationPlan = new RelationPlanner(analysis, symbolAllocator, idAllocator, lambdaDeclarationToSymbolMap, metadata, outerContext, session, recursiveSubqueries)
+                .process(table, null);
+
+        PlanBuilder builder = newPlanBuilder(relationPlan, analysis, lambdaDeclarationToSymbolMap);
+
+        if (node.getWhere().isPresent()) {
+            builder = filter(builder, node.getWhere().get(), node);
+        }
+
+        builder = builder.appendProjections(orderedExpressions, symbolAllocator, idAllocator);
+
+        PlanAndMappings planAndMappings = coerce(builder, orderedExpressions, analysis, idAllocator, symbolAllocator, typeCoercion);
+        builder = planAndMappings.getSubPlan();
+        ProjectNode projectNode = (ProjectNode) builder.getRoot();
+
+        List<Symbol> expressionSymbols = orderedExpressions.stream()
+                .map(expression -> requireNonNull(planAndMappings.get(expression), "planAndMappings.get(expression) is null"))
+                .collect(Collectors.toList());
+
+        Set<String> dependencyColumnNames = analysis.getColumnReferences().stream()
+                .map(NodeRef::getNode)
+                .filter(someNode -> someNode instanceof Identifier)
+                .map(someNode -> ((Identifier) someNode).getValue())
+                .collect(toImmutableSet());
+        List<ColumnHandle> dependencyColumns = dataColumnHandles.entrySet().stream()
+                .filter(entry -> dependencyColumnNames.contains(entry.getKey()))
+                .map(Map.Entry::getValue)
+                .collect(toImmutableList());
+
+        Symbol rowId = builder.translate(analysis.getRowIdField(table));
+
+        expressionSymbols.add(rowId);
+        List<Integer> updateExpressionChannelNumbers = createUpdateExpressionChannelNumbers(projectNode.getAssignments(), expressionSymbols);
+
+        List<ColumnHandle> targets = dataColumnHandles.entrySet().stream()
+                .filter(entry -> targetColumnNames.contains(entry.getKey()))
+                .map(Map.Entry::getValue)
+                .collect(toImmutableList());
+
+        List<Symbol> projectedSymbols = new ArrayList<>();
+        projectedSymbols.add(rowId);
+        projectedSymbols.addAll(planAndMappings.mappings.values());
+
+        List<Symbol> outputs = ImmutableList.of(
+                symbolAllocator.newSymbol("partialrows", BIGINT),
+                symbolAllocator.newSymbol("fragment", VARBINARY));
+
+        // create update node
+        return new UpdateNode(
+                idAllocator.getNextId(),
+                builder.getRoot(),
+                new UpdateTarget(handle, metadata.getTableMetadata(session, handle).getTable(), targets, dependencyColumns, updateExpressionChannelNumbers),
+                rowId,
+                projectedSymbols,
```

Comment:
What's `projectedSymbols`

Reply:
projectedSymbols are the symbols for the UPDATE SET expressions, plus the rowId symbol.

Reply:
I renamed `projectedSymbols` to be `expressionSymbols`.